### ISSUE TITLE
How to test the correctness of the data?

### ISSUE THREAD

**Issue by adborden** [caciviclab/disclosure-backend-static#132]
Conversation about our testing strategy.

**Comment by adborden** [comment]
## Tiered approach

There's several steps in our ETL, from downloading to data cleaning to import, to hepler views, to calculations for different entities. We could implement assertions at each level (tier) in the ETL process to catch errors early.

I'm not sure if we have a sense of at what level our errors are coming from. I supsect it's in the last step of calculations, which is the bulk of the process. I wonder if we can split it into smaller chucks that are simpler and easier to make assertions against.

**Comment by sfdoran** [comment]
Thanks @adborden for starting this conversation. 

One place we could create an automated comparison would be to compare the total when each filing period is summed against the year-to-date total on the most recent filing plus any subsequent 24-hour filings.

Once source for differences may be that OpenDisclosure sums the total contributions and total expenditures for each filing period. If a committee edits prior transactions and then doesn't amend the prior filings those totals won't be updated in the data for the earlier periods, but it will affect the overall total reported on subsequent filings.

On filer error, we do have a disclaimer in our FAQ. But some problems aren't a genuine filer error, but rather difficulties caused by the way the data is reported. I think a very generic statement is better with a link to our note stating we don't clean the data.

**Comment by tdooner** [comment]
Agreed, there are a lot of different things we can/should be testing here.

The first tests, which I just merged yesterday in #127 , seek to ensure that our calculations match our mental models of what is happening. We do this by creating trivial, static, test cases that are much easier to reason about. This is purely to test that our code does what we think it does.

I'd like to propose that we create a distinction between "tests" and filer QA scripts that can detect filer error.

For the "tests" that we write, the goal should be that for any individual change to OpenDisclosure, all of the tests are passing, which indicates that there are no unexpected changes in behavior.

For filer QA, we can do whatever we want. If there are particular things we want to check, and report on them somehow, sure, let's do that. But it's not in our power to fix QA issues, so I'd love to separate them conceptually from "tests".

Right now, IMHO, our priority should be to add more test cases, especially as we hit buggy times of the filing season (i.e. when 496/497 data needs to be de-duplicated). When we get a good set of tests that make sure that the numbers we're calculating for ballot measures and candidates are right, then that should free up Suzanne from having to manually check all the numbers all the time. That'd be nice, wouldn't it? 😎

**Comment by adborden** [comment]
Quoting Tom Dooner (2018-07-30 17:08:46)
> I'd like to propose that we create a distinction between "tests" and filer QA scripts that can detect filer error.

Great idea!

--
Aaron D Borden
Human and hacker

**Comment by tdooner** [comment]
QA wishlist:
- [ ] List of Ballot Measure Committees (and which ballot measures they contribute to) that aren't on the spreadsheet

**Comment by sfdoran** [comment]
@tdooner I can get those lists together for you. Just committees active in the 2016 election and forward, right?
